File: src/legal_format_engine/renderers/docx_renderer.py
"""DOCX renderer for legal documents.

Produces properly formatted Word documents from the internal
LegalDocument representation using python-docx.
"""

from __future__ import annotations

from pathlib import Path

from docx import Document as DocxDocument
from docx.enum.section import WD_ORIENT
from docx.enum.text import WD_ALIGN_PARAGRAPH, WD_LINE_SPACING
from docx.oxml import OxmlElement
from docx.oxml.ns import qn
from docx.shared import Inches, Pt, RGBColor

from legal_format_engine.models.document import (
    Alignment,
    CaptionBlock,
    ContentBlock,
    LegalDocument,
    Section,
    SignatureBlock,
)
from legal_format_engine.models.letterhead import Letterhead, LetterheadLine
from legal_format_engine.rules.schema import PageFormat, Ruleset

# Alignment mapping
_ALIGN_MAP = {
    Alignment.LEFT: WD_ALIGN_PARAGRAPH.LEFT,
    Alignment.CENTER: WD_ALIGN_PARAGRAPH.CENTER,
    Alignment.RIGHT: WD_ALIGN_PARAGRAPH.RIGHT,
    Alignment.JUSTIFY: WD_ALIGN_PARAGRAPH.JUSTIFY,
}


def render_docx(
    doc: LegalDocument,
    ruleset: Ruleset,
    output_path: str | Path,
    letterhead: Letterhead | None = None,
) -> Path:
    """Render a LegalDocument to a DOCX file."""
    output_path = Path(output_path)
    docx = DocxDocument()
    fmt = ruleset.page_format

    _setup_page(docx, fmt)
    _set_default_font(docx, fmt)
    _enable_hyphenation(docx)

    if letterhead:
        _render_letterhead(docx, letterhead, fmt)

    if doc.caption:
        _render_caption(docx, doc.caption, fmt)
        # Page break after caption — body starts on a new page
        docx.add_page_break()

    for section in doc.sections:
        _render_section(docx, section, fmt, ruleset)

    if doc.signature_block:
        _add_blank_line(docx, fmt)
        _render_signature_block(docx, doc.signature_block, fmt)

    for cert in doc.certifications:
        docx.add_page_break()
        _render_section(docx, cert, fmt, ruleset)

    docx.save(str(output_path))
    return output_path


def _setup_page(docx: DocxDocument, fmt: PageFormat) -> None:
    """Configure page dimensions and margins."""
    section = docx.sections[0]
    section.page_width = Inches(fmt.page_width_inches)
    section.page_height = Inches(fmt.page_height_inches)
    section.top_margin = Inches(fmt.margin_top_inches)
    section.bottom_margin = Inches(fmt.margin_bottom_inches)
    section.left_margin = Inches(fmt.margin_left_inches)
    section.right_margin = Inches(fmt.margin_right_inches)
    section.orientation = WD_ORIENT.PORTRAIT


def _set_default_font(docx: DocxDocument, fmt: PageFormat) -> None:
    """Set the default font for the document."""
    style = docx.styles["Normal"]
    font = style.font
    font.name = fmt.font_name
    font.size = Pt(fmt.font_size_pt)
    pf = style.paragraph_format
    pf.line_spacing_rule = WD_LINE_SPACING.DOUBLE
    pf.space_after = Pt(0)
    pf.space_before = Pt(0)


def _enable_hyphenation(docx: DocxDocument) -> None:
    """Enable automatic hyphenation to prevent excessive word spacing with JUSTIFY."""
    settings = docx.settings.element
    auto_hyphen = OxmlElement("w:autoHyphenation")
    auto_hyphen.set(qn("w:val"), "true")
    settings.append(auto_hyphen)


# ── Caption ──────────────────────────────────────────────────────────


def _render_caption(docx: DocxDocument, caption: CaptionBlock, fmt: PageFormat) -> None:
    """Render the caption block with single spacing."""
    for line in caption.lines:
        para = docx.add_paragraph()
        alignment = _ALIGN_MAP.get(line.alignment, WD_ALIGN_PARAGRAPH.CENTER)
        para.alignment = alignment

        # Caption lines are single-spaced
        pf = para.paragraph_format
        pf.line_spacing_rule = WD_LINE_SPACING.SINGLE
        pf.space_before = Pt(0)
        pf.space_after = Pt(2)

        run = para.add_run(line.text)
        run.font.name = fmt.font_name
        run.font.size = Pt(fmt.font_size_pt)
        run.bold = line.bold
        run.italic = line.italic
        run.underline = line.underline


# ── Sections ─────────────────────────────────────────────────────────


def _render_section(
    docx: DocxDocument,
    section: Section,
    fmt: PageFormat,
    ruleset: Ruleset,
) -> None:
    """Render a section with its heading and content."""
    if section.heading_text:
        heading_rule = ruleset.get_heading_rule(section.heading_level)
        heading_bold = heading_rule.bold if heading_rule else True

        # Build full heading text with prefix
        heading_text = section.heading_text
        if section.numbering_prefix:
            heading_text = f"{section.numbering_prefix} {heading_text}"

        alignment = Alignment.CENTER
        if heading_rule:
            alignment = Alignment(heading_rule.alignment)

        # Determine heading indent based on heading level
        indent = 0.0
        if heading_rule:
            indent = heading_rule.indent_inches

        _add_heading_paragraph(docx, heading_text, fmt, alignment,
                               heading_bold, indent)

    for block in section.content:
        _add_content_paragraph(docx, block, fmt)

    for subsection in section.subsections:
        _render_section(docx, subsection, fmt, ruleset)


def _add_heading_paragraph(
    docx: DocxDocument,
    text: str,
    fmt: PageFormat,
    alignment: Alignment,
    bold: bool,
    indent_inches: float = 0.0,
) -> None:
    """Add a heading paragraph (no first-line indent, double-spaced)."""
    para = docx.add_paragraph()
    para.alignment = _ALIGN_MAP.get(alignment, WD_ALIGN_PARAGRAPH.LEFT)

    pf = para.paragraph_format
    pf.line_spacing_rule = WD_LINE_SPACING.DOUBLE
    pf.space_before = Pt(0)
    pf.space_after = Pt(0)
    pf.first_line_indent = Inches(0)  # headings never have first-line indent
    if indent_inches > 0:
        pf.left_indent = Inches(indent_inches)

    run = para.add_run(text)
    run.font.name = fmt.font_name
    run.font.size = Pt(fmt.font_size_pt)
    run.bold = bold


def _add_content_paragraph(
    docx: DocxDocument,
    block: ContentBlock,
    fmt: PageFormat,
) -> None:
    """Add a single content block as a paragraph.

    Body text gets first-line indent and justified alignment.
    Caption text gets single spacing and no indent.
    """
    para = docx.add_paragraph()

    # Determine alignment
    if block.alignment:
        para.alignment = _ALIGN_MAP.get(block.alignment, WD_ALIGN_PARAGRAPH.JUSTIFY)
    else:
        para.alignment = WD_ALIGN_PARAGRAPH.JUSTIFY

    pf = para.paragraph_format

    if block.is_caption:
        # Caption lines: single-spaced, no indent
        pf.line_spacing_rule = WD_LINE_SPACING.SINGLE
        pf.space_before = Pt(0)
        pf.space_after = Pt(2)
        pf.first_line_indent = Inches(0)
    elif block.is_body_text:
        # Body text: double-spaced, first-line indent, justified
        pf.line_spacing_rule = WD_LINE_SPACING.DOUBLE
        pf.space_before = Pt(0)
        pf.space_after = Pt(0)
        pf.first_line_indent = Inches(fmt.first_line_indent_inches)
    else:
        # Default: double-spaced, no indent (headings, signature lines, etc.)
        pf.line_spacing_rule = WD_LINE_SPACING.DOUBLE
        pf.space_before = Pt(0)
        pf.space_after = Pt(0)
        pf.first_line_indent = Inches(0)

    # Apply any explicit left indent
    if block.indent_inches > 0:
        pf.left_indent = Inches(block.indent_inches)

    run = para.add_run(block.text)
    run.font.name = fmt.font_name
    run.font.size = Pt(fmt.font_size_pt)
    run.bold = block.bold
    run.italic = block.italic
    run.underline = block.underline


def _add_blank_line(docx: DocxDocument, fmt: PageFormat) -> None:
    """Add a blank paragraph as a spacer."""
    para = docx.add_paragraph()
    run = para.add_run("")
    run.font.name = fmt.font_name
    run.font.size = Pt(fmt.font_size_pt)


# ── Signature block ──────────────────────────────────────────────────


def _render_signature_block(
    docx: DocxDocument,
    sig: SignatureBlock,
    fmt: PageFormat,
) -> None:
    """Render the attorney signature block."""
    lines = [
        "Respectfully submitted,",
        "",
        "____________________________",
        sig.attorney_name,
        f"State Bar No. {sig.bar_number}",
    ]
    if sig.firm:
        lines.append(sig.firm)
    lines.extend([
        sig.address,
        f"Phone: {sig.phone}",
        f"Email: {sig.email}",
    ])

    for text in lines:
        block = ContentBlock(text=text, alignment=Alignment.LEFT)
        _add_content_paragraph(docx, block, fmt)


# ── Letterhead ───────────────────────────────────────────────────────

_LETTERHEAD_ALIGN = {
    "left": WD_ALIGN_PARAGRAPH.LEFT,
    "center": WD_ALIGN_PARAGRAPH.CENTER,
    "right": WD_ALIGN_PARAGRAPH.RIGHT,
}


def _render_letterhead(docx: DocxDocument, lh: Letterhead, fmt: PageFormat) -> None:
    """Render a letterhead at the top of the document."""
    # Logo (if present)
    if lh.logo_path:
        logo = Path(lh.logo_path)
        if logo.exists():
            para = docx.add_paragraph()
            para.alignment = WD_ALIGN_PARAGRAPH.CENTER
            run = para.add_run()
            run.add_picture(str(logo), width=Inches(lh.logo_width_inches))

    # Text lines
    for line in lh.lines:
        _render_letterhead_line(docx, line, fmt)

    # Separator
    if lh.separator_line:
        para = docx.add_paragraph()
        para.alignment = WD_ALIGN_PARAGRAPH.CENTER
        # Use a bottom border on the paragraph
        pf = para.paragraph_format
        pf.space_before = Pt(4)
        pf.space_after = Pt(lh.spacing_after_pt)
        run = para.add_run("_" * 60)
        run.font.size = Pt(6)
        run.font.color.rgb = RGBColor(0x99, 0x99, 0x99)
    else:
        # Just add spacing
        para = docx.add_paragraph()
        para.paragraph_format.space_after = Pt(lh.spacing_after_pt)


def _render_letterhead_line(
    docx: DocxDocument, line: LetterheadLine, fmt: PageFormat
) -> None:
    """Render a single letterhead line."""
    para = docx.add_paragraph()
    para.alignment = _LETTERHEAD_ALIGN.get(line.alignment, WD_ALIGN_PARAGRAPH.CENTER)

    # Tight spacing for letterhead
    pf = para.paragraph_format
    pf.space_before = Pt(0)
    pf.space_after = Pt(0)
    pf.line_spacing_rule = WD_LINE_SPACING.SINGLE

    run = para.add_run(line.text)
    run.font.name = fmt.font_name
    run.font.size = Pt(line.font_size_pt or fmt.font_size_pt)
    run.bold = line.bold
    run.italic = line.italic

File: legal-compliance-engine-pkg/src/LegalEngine.Rendering/DocxRenderer.fs
namespace LegalEngine.Rendering

open System
open System.IO
open DocumentFormat.OpenXml
open DocumentFormat.OpenXml.Packaging
open DocumentFormat.OpenXml.Wordprocessing
open LegalEngine.Domain

/// DOCX renderer for legal documents.
/// Produces properly formatted Word documents from the internal
/// LegalDocument representation using DocumentFormat.OpenXml.
module DocxRenderer =

    // ── Conversion helpers ──────────────────────────────────────────

    /// Convert inches to EMUs (English Metric Units). 1 inch = 914400 EMUs.
    let private inchesToEmu (inches: float) : int64 =
        int64 (inches * 914400.0)

    /// Convert inches to twentieths of a point (twips). 1 inch = 1440 twips.
    let private inchesToTwips (inches: float) : int =
        int (inches * 1440.0)

    /// Convert points to half-points. OpenXml font sizes are in half-points.
    let private ptToHalfPoints (pt: float) : string =
        string (int (pt * 2.0))

    /// Convert points to twentieths of a point.
    let private ptToTwips (pt: float) : string =
        string (int (pt * 20.0))

    // ── Alignment mapping ───────────────────────────────────────────

    let private toJustification (alignment: Alignment) : JustificationValues =
        match alignment with
        | Alignment.Left -> JustificationValues.Left
        | Alignment.Center -> JustificationValues.Center
        | Alignment.Right -> JustificationValues.Right
        | Alignment.Justify -> JustificationValues.Both

    let private alignmentFromString (s: string) : JustificationValues =
        match s.ToLowerInvariant() with
        | "left" -> JustificationValues.Left
        | "center" | "centre" -> JustificationValues.Center
        | "right" -> JustificationValues.Right
        | "justify" | "both" -> JustificationValues.Both
        | _ -> JustificationValues.Both

    // ── Page setup ──────────────────────────────────────────────────

    let private setupPage (body: Body) (fmt: PageFormat) =
        let sectPr = SectionProperties()
        let pgSz = PageSize()
        pgSz.Width <- UInt32Value(uint32 (inchesToTwips fmt.PageWidthInches))
        pgSz.Height <- UInt32Value(uint32 (inchesToTwips fmt.PageHeightInches))
        pgSz.Orient <- EnumValue(PageOrientationValues.Portrait)
        sectPr.AppendChild(pgSz) |> ignore

        let pgMar = PageMargin()
        pgMar.Top <- Int32Value(int32 (inchesToTwips fmt.MarginTopInches))
        pgMar.Bottom <- Int32Value(int32 (inchesToTwips fmt.MarginBottomInches))
        pgMar.Left <- UInt32Value(uint32 (inchesToTwips fmt.MarginLeftInches))
        pgMar.Right <- UInt32Value(uint32 (inchesToTwips fmt.MarginRightInches))
        sectPr.AppendChild(pgMar) |> ignore

        body.AppendChild(sectPr) |> ignore

    // ── Default font via style ──────────────────────────────────────

    let private setDefaultFont (stylesPart: StyleDefinitionsPart) (fmt: PageFormat) =
        let styles = Styles()

        // Default run properties
        let docDefaults = DocDefaults()
        let runPrDefault = RunPropertiesDefault()
        let rPr = RunPropertiesBaseStyle()
        let runFonts = RunFonts()
        runFonts.Ascii <- StringValue(fmt.FontName)
        runFonts.HighAnsi <- StringValue(fmt.FontName)
        rPr.AppendChild(runFonts) |> ignore
        let fontSize = FontSize()
        fontSize.Val <- StringValue(ptToHalfPoints fmt.FontSizePt)
        rPr.AppendChild(fontSize) |> ignore
        runPrDefault.RunPropertiesBaseStyle <- rPr
        docDefaults.RunPropertiesDefault <- runPrDefault

        // Default paragraph properties
        let parPrDefault = ParagraphPropertiesDefault()
        let pPr = ParagraphPropertiesBaseStyle()
        let spacing = SpacingBetweenLines()
        // Double spacing: 480 twips (24pt line = 2x 12pt)
        spacing.Line <- StringValue(string (int (fmt.LineSpacing * 240.0)))
        spacing.LineRule <- EnumValue(LineSpacingRuleValues.Auto)
        spacing.Before <- StringValue("0")
        spacing.After <- StringValue("0")
        pPr.AppendChild(spacing) |> ignore
        parPrDefault.ParagraphPropertiesBaseStyle <- pPr
        docDefaults.ParagraphPropertiesDefault <- parPrDefault

        styles.AppendChild(docDefaults) |> ignore
        stylesPart.Styles <- styles

    // ── Hyphenation ─────────────────────────────────────────────────

    let private enableHyphenation (settingsPart: DocumentSettingsPart) =
        let settings = Settings()
        let autoHyphen = AutoHyphenation()
        autoHyphen.Val <- OnOffValue(true)
        settings.AppendChild(autoHyphen) |> ignore
        settingsPart.Settings <- settings

    // ── Run creation ────────────────────────────────────────────────

    let private makeRun (text: string) (fmt: PageFormat) (bold: bool) (italic: bool) (underline: bool) : Run =
        let run = Run()
        let rPr = RunProperties()

        let runFonts = RunFonts()
        runFonts.Ascii <- StringValue(fmt.FontName)
        runFonts.HighAnsi <- StringValue(fmt.FontName)
        rPr.AppendChild(runFonts) |> ignore

        let fontSize = FontSize()
        fontSize.Val <- StringValue(ptToHalfPoints fmt.FontSizePt)
        rPr.AppendChild(fontSize) |> ignore

        if bold then
            rPr.AppendChild(Bold()) |> ignore
        if italic then
            rPr.AppendChild(Italic()) |> ignore
        if underline then
            let u = Underline()
            u.Val <- EnumValue(UnderlineValues.Single)
            rPr.AppendChild(u) |> ignore

        run.RunProperties <- rPr
        let t = Text(text)
        t.Space <- EnumValue(SpaceProcessingModeValues.Preserve)
        run.AppendChild(t) |> ignore
        run

    // ── Paragraph creation helpers ──────────────────────────────────

    let private makeParagraph
        (text: string)
        (fmt: PageFormat)
        (justification: JustificationValues)
        (lineSpacingTwips: int)
        (firstLineIndentTwips: int)
        (leftIndentTwips: int)
        (bold: bool)
        (italic: bool)
        (underline: bool)
        (spaceAfterPt: float) : Paragraph =

        let para = Paragraph()
        let pPr = ParagraphProperties()
        let jc = Justification()
        jc.Val <- EnumValue(justification)
        pPr.AppendChild(jc) |> ignore

        let spacing = SpacingBetweenLines()
        spacing.Line <- StringValue(string lineSpacingTwips)
        spacing.LineRule <- EnumValue(LineSpacingRuleValues.Auto)
        spacing.Before <- StringValue("0")
        spacing.After <- StringValue(ptToTwips spaceAfterPt)
        pPr.AppendChild(spacing) |> ignore

        if firstLineIndentTwips > 0 || leftIndentTwips > 0 then
            let ind = Indentation()
            if firstLineIndentTwips > 0 then
                ind.FirstLine <- StringValue(string firstLineIndentTwips)
            if leftIndentTwips > 0 then
                ind.Left <- StringValue(string leftIndentTwips)
            pPr.AppendChild(ind) |> ignore

        para.ParagraphProperties <- pPr
        let run = makeRun text fmt bold italic underline
        para.AppendChild(run) |> ignore
        para

    // ── Caption rendering ───────────────────────────────────────────

    let private renderCaption (body: Body) (caption: CaptionBlock) (fmt: PageFormat) =
        for line in caption.Lines do
            let justification = toJustification line.Alignment
            // Single spacing = 240 twips
            let para =
                makeParagraph
                    line.Text fmt justification
                    240 0 0
                    line.Bold line.Italic line.Underline
                    2.0
            body.AppendChild(para) |> ignore

    // ── Page break ──────────────────────────────────────────────────

    let private addPageBreak (body: Body) =
        let para = Paragraph()
        let run = Run()
        let br = Break()
        br.Type <- EnumValue(BreakValues.Page)
        run.AppendChild(br) |> ignore
        para.AppendChild(run) |> ignore
        body.AppendChild(para) |> ignore

    // ── Section rendering ───────────────────────────────────────────

    let private renderHeading (body: Body) (text: string) (fmt: PageFormat) (ruleset: Ruleset) (level: HeadingLevel) (numberingPrefix: string option) =
        let headingRule = Ruleset.getHeadingRule (HeadingLevel.toInt level) ruleset
        let bold =
            match headingRule with
            | Some r -> r.Bold
            | None -> true

        let fullText =
            match numberingPrefix with
            | Some prefix -> prefix + " " + text
            | None -> text

        let justification =
            match headingRule with
            | Some r -> alignmentFromString r.Alignment
            | None -> JustificationValues.Center

        let leftIndentTwips =
            match headingRule with
            | Some r -> inchesToTwips r.IndentInches
            | None -> 0

        // Double spacing = 480 twips; headings never have first-line indent
        let para =
            makeParagraph
                fullText fmt justification
                480 0 leftIndentTwips
                bold false false
                0.0
        body.AppendChild(para) |> ignore

    let private renderContentBlock (body: Body) (block: ContentBlock) (fmt: PageFormat) =
        let justification = toJustification block.Alignment

        let lineSpacing, spaceAfter, firstLineIndent =
            if block.IsCaption then
                // Caption: single-spaced, no indent
                240, 2.0, 0
            elif block.IsBodyText then
                // Body: double-spaced, first-line indent, justified
                480, 0.0, inchesToTwips fmt.FirstLineIndentInches
            else
                // Default: double-spaced, no indent
                480, 0.0, 0

        let leftIndentTwips =
            if block.IndentInches > 0.0 then inchesToTwips block.IndentInches
            else 0

        let para =
            makeParagraph
                block.Text fmt justification
                lineSpacing firstLineIndent leftIndentTwips
                block.Bold block.Italic block.Underline
                spaceAfter
        body.AppendChild(para) |> ignore

    let rec private renderSection (body: Body) (section: Section) (fmt: PageFormat) (ruleset: Ruleset) =
        if not (String.IsNullOrWhiteSpace section.HeadingText) then
            renderHeading body section.HeadingText fmt ruleset section.HeadingLevel section.NumberingPrefix

        for block in section.Content do
            renderContentBlock body block fmt

        for sub in section.Subsections do
            renderSection body sub fmt ruleset

    // ── Blank line ──────────────────────────────────────────────────

    let private addBlankLine (body: Body) (fmt: PageFormat) =
        let para = makeParagraph "" fmt JustificationValues.Left 480 0 0 false false false 0.0
        body.AppendChild(para) |> ignore

    // ── Signature block ─────────────────────────────────────────────

    let private renderSignatureBlock (body: Body) (sig': SignatureBlockInfo) (fmt: PageFormat) =
        let lines = [
            "Respectfully submitted,"
            ""
            "____________________________"
            sig'.AttorneyName
            sprintf "State Bar No. %s" sig'.BarNumber
        ]
        let lines =
            match sig'.Firm with
            | Some firm -> lines @ [ firm ]
            | None -> lines
        let lines =
            lines @ [
                sig'.Address
                sprintf "Phone: %s" sig'.Phone
                sprintf "Email: %s" sig'.Email
            ]

        for text in lines do
            let para =
                makeParagraph
                    text fmt JustificationValues.Left
                    480 0 0
                    false false false
                    0.0
            body.AppendChild(para) |> ignore

    // ── Public API ──────────────────────────────────────────────────

    /// Render a LegalDocument to a DOCX file at the specified output path.
    let renderDocx (doc: LegalDocument) (ruleset: Ruleset) (outputPath: string) : unit =
        use wordDoc = WordprocessingDocument.Create(outputPath, WordprocessingDocumentType.Document)

        let mainPart = wordDoc.AddMainDocumentPart()
        let document = Document()
        let body = Body()

        // Set up styles (default font, spacing)
        let stylesPart = mainPart.AddNewPart<StyleDefinitionsPart>()
        setDefaultFont stylesPart ruleset.PageFormat

        // Enable hyphenation
        let settingsPart = mainPart.AddNewPart<DocumentSettingsPart>()
        enableHyphenation settingsPart

        // Render caption
        match doc.Caption with
        | Some caption ->
            renderCaption body caption ruleset.PageFormat
            addPageBreak body
        | None -> ()

        // Render sections
        for section in doc.Sections do
            renderSection body section ruleset.PageFormat ruleset

        // Render signature block
        match doc.SignatureBlock with
        | Some sig' ->
            addBlankLine body ruleset.PageFormat
            renderSignatureBlock body sig' ruleset.PageFormat
        | None -> ()

        // Render certifications with page breaks between
        for cert in doc.Certifications do
            addPageBreak body
            renderSection body cert ruleset.PageFormat ruleset

        // Set up page format (margins, page size) via section properties
        setupPage body ruleset.PageFormat

        document.AppendChild(body) |> ignore
        mainPart.Document <- document
